File: Server~/src/tools/assetTools/addAssetTool.ts
import * as z from 'zod';
import { Logger } from '../../utils/logger.js';
import { McpUnity } from '../../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'add_asset_to_project';
const toolDescription = 'Imports assets (images, etc.) into the Unity project';
const paramsSchema = z.object({
  sourcePaths: z.union([
    z.string(),
    z.array(z.string())
  ]).describe('The source path(s) of the asset(s) to import (file not directory)'),
  destPath: z.string().describe('The destination path where the asset(s) will be imported to')
});

export function registerAddAssetTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
      
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const { sourcePaths, destPath } = params;
  
  // Convert single path to array if needed
  const paths = Array.isArray(sourcePaths) ? sourcePaths : [sourcePaths];
        
  if (!paths || !destPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      'Required parameters "sourcePaths" and "destPath" must be provided'
    );
  }

  if (paths.length === 0) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      'At least one source path must be provided'
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      sourcePaths: paths,
      destPath
    }
  });
  
  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to import assets to ${destPath}`,
      response.errors
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message,
    }]
  };
}

File: Editor/Tools/AssetTools/AddAssetTool.cs
using UnityEngine;
using UnityEditor;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using McpUnity.Tools;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    public class AddAssetTool : McpToolBase
    {
        public AddAssetTool()
        {
            Name = "add_asset_to_project";
            Description = "Imports assets (images, etc.) into the Unity project";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            JToken sourcePathsToken = parameters["sourcePaths"];
            string destPath = parameters["destPath"]?.ToObject<string>();

            if (sourcePathsToken == null || string.IsNullOrEmpty(destPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameters 'sourcePaths' and 'destPath' must be provided",
                    "validation_error"
                );
            }

            List<string> sourcePaths;
            if (sourcePathsToken.Type == JTokenType.Array)
            {
                sourcePaths = sourcePathsToken.ToObject<List<string>>();
            }
            else
            {
                sourcePaths = new List<string> { sourcePathsToken.ToObject<string>() };
            }

            if (!sourcePaths.Any())
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "At least one source path must be provided",
                    "validation_error"
                );
            }

            destPath = AssetUtils.EnsureAssetPath(destPath);

            int results = 0;
            Dictionary<string, string> errors = new Dictionary<string, string>();

            foreach (string sourcePath in sourcePaths)
            {
                if (!File.Exists(sourcePath))
                {
                    if (Directory.Exists(sourcePath))
                    {
                        errors.Add(sourcePath, "Path must be a file, not a directory");
                    } else
                    {
                        errors.Add(sourcePath, "Source file not found");
                    }
                    continue;
                }

                string fileName = Path.GetFileName(sourcePath);
                string finalDestPath = Path.Combine(destPath, fileName);
                string normalizedDestPath = AssetUtils.EnsureAssetPath(finalDestPath);

                if (AssetUtils.CheckAssetExists(normalizedDestPath))
                {
                    errors.Add(finalDestPath, "Asset already exists at destination");
                    continue;
                }

                try
                {
                    AssetUtils.EnsureDirectoryExists(Path.GetDirectoryName(normalizedDestPath));

                    File.Copy(sourcePath, normalizedDestPath, false);

                    AssetDatabase.ImportAsset(normalizedDestPath, ImportAssetOptions.ForceUpdate | ImportAssetOptions.ForceSynchronousImport);
                    ++results;

                    AssetDatabase.Refresh();
                }
                catch (System.Exception ex)
                {
                    if (File.Exists(normalizedDestPath))
                    {
                        File.Delete(normalizedDestPath);
                    }
                    errors.Add(sourcePath, ex.Message);
                }
            }

            return new JObject
            {
                ["success"] = !errors.Any(),
                ["type"] = "text",
                ["message"] = $"Imported assets: {results} successfully, {errors.Count} failed",
                ["errors"] = errors.Count > 0 ? JObject.FromObject(errors) : null
            };
        }
    }
}
